bug: provider cleanup can fail after sandbox deletion conflict

## Agent Diagnostic

Skills loaded: `watch-github-actions`, `create-github-issue`.

Investigation method: Watched PR #1516 checks after applying the `test:e2e` label, inspected the failed Python E2E job through the GitHub Actions logs, and compared the failure with the Docker routing change in the PR.

Findings:

- PR: https://github.com/NVIDIA/OpenShell/pull/1516
- Workflow run: https://github.com/NVIDIA/OpenShell/actions/runs/26269734764
- Failed job: https://github.com/NVIDIA/OpenShell/actions/runs/26269734764/job/77321344927
- `e2e / E2E (rust-docker)` passed.
- `e2e / E2E (rust-podman)` passed.
- `e2e / E2E (python)` failed in provider cleanup, not during gateway startup or Docker callback routing.
- The failing Python test created and exercised a sandbox, then cleanup hit an optimistic-concurrency error while moving the sandbox to `Deleting`. Provider cleanup then failed because the sandbox was still attached.

## Description

**Actual behavior:** Python E2E can fail during cleanup when deleting a sandbox races with another sandbox resource update. The sandbox deletion RPC returns `ABORTED`, and the provider cleanup then fails with `FAILED_PRECONDITION` because the provider is still attached to the sandbox.

**Expected behavior:** Cleanup paths should be resilient to transient sandbox resource-version conflicts. A delete request should retry the phase transition or otherwise leave the sandbox/provider relationship in a cleanup-safe state so provider deletion does not fail after a transient concurrent modification.

## Reproduction Steps

1. Run the Python E2E suite with Docker-backed sandboxes on Linux ARM64, as in the `e2e / E2E (python)` GitHub Actions job.
2. Observe `e2e/python/test_sandbox_providers.py::test_nvidia_provider_injects_nvidia_api_key_env_var`.
3. The test can fail during cleanup after sandbox creation and execution have succeeded.

## Environment

- OS: Linux ARM64 GitHub Actions runner (`linux-arm64-cpu8`)
- Docker API: 1.54
- OpenShell commit: `c0a9306b1884d4584dc36b9f0b9cc85942da3dad`
- Workflow: `Branch E2E Checks`
- Job: `e2e / E2E (python)`

## Logs

```text
FAILED e2e/python/test_sandbox_providers.py::test_nvidia_provider_injects_nvidia_api_key_env_var

grpc._channel._InactiveRpcError:
  status = StatusCode.ABORTED
  details = "set sandbox phase to Deleting failed due to concurrent modification (current resource_version: 3)"

During handling of the above exception, another exception occurred:

grpc._channel._InactiveRpcError:
  status = StatusCode.FAILED_PRECONDITION
  details = "provider 'e2e-test-nvidia-provider-env' is attached to sandbox(es): ecstatic-bug"

1 failed, 75 passed in 54.98s
```

--- a/crates/openshell-server/src/compute/mod.rs
+++ b/crates/openshell-server/src/compute/mod.rs
@@ -51,6 +51,8 @@ type DriverWatchStream = Pin<Box<dyn Stream<Item = Result<WatchSandboxesEvent, S
 type SharedComputeDriver =
     Arc<dyn ComputeDriver<WatchSandboxesStream = DriverWatchStream> + Send + Sync>;
 
+const DELETE_PHASE_CAS_RETRY_LIMIT: usize = 3;
+
 #[tonic::async_trait]
 trait ShutdownCleanup: Send + Sync {
     async fn cleanup_on_shutdown(&self) -> Result<(), String>;
@@ -540,17 +542,7 @@ impl ComputeRuntime {
 
         let id = sandbox.object_id().to_string();
 
-        // Use CAS to set phase to Deleting
-        // TODO: Accept expected_version from DeleteSandboxRequest for proper client-driven CAS
-        let sandbox = self
-            .store
-            .update_message_cas::<Sandbox, _>(&id, 0, |s| {
-                s.set_phase(SandboxPhase::Deleting as i32);
-            })
-            .await
-            .map_err(|e| {
-                crate::grpc::persistence_error_to_status(e, "set sandbox phase to Deleting")
-            })?;
+        let sandbox = self.set_sandbox_phase_deleting_with_retry(&id).await?;
 
         self.sandbox_index.update_from_sandbox(&sandbox);
         self.sandbox_watch_bus.notify(&id);
@@ -572,6 +564,114 @@ impl ComputeRuntime {
 
         self.cleanup_sandbox_state(&id);
         Ok(deleted)
+    }
+
+    async fn set_sandbox_phase_deleting_with_retry(
+        &self,
+        sandbox_id: &str,
+    ) -> Result<Sandbox, Status> {
+        self.set_sandbox_phase_deleting_with_initial_snapshot(sandbox_id, None)
+            .await
+    }
+
+    async fn set_sandbox_phase_deleting_with_initial_snapshot(
+        &self,
+        sandbox_id: &str,
+        mut initial_snapshot: Option<Sandbox>,
+    ) -> Result<Sandbox, Status> {
+        let operation = "set sandbox phase to Deleting";
+
+        for attempt in 1..=DELETE_PHASE_CAS_RETRY_LIMIT {
+            let sandbox = match initial_snapshot.take() {
+                Some(sandbox) => sandbox,
+                None => self
+                    .store
+                    .get_message::<Sandbox>(sandbox_id)
+                    .await
+                    .map_err(|e| Status::internal(format!("fetch sandbox failed: {e}")))?
+                    .ok_or_else(|| Status::not_found("sandbox not found"))?,
+            };
+
+            match self
+                .write_sandbox_phase_deleting_from_snapshot(sandbox)
+                .await
+            {
+                Ok(sandbox) => {
+                    if attempt > 1 {
+                        debug!(
+                            sandbox_id,
+                            attempt, "Retried sandbox delete phase transition after CAS conflict"
+                        );
+                    }
+                    return Ok(sandbox);
+                }
+                Err(crate::persistence::PersistenceError::Conflict {
+                    current_resource_version,
+                }) => {
+                    let err = crate::persistence::PersistenceError::Conflict {
+                        current_resource_version,
+                    };
+                    if attempt == DELETE_PHASE_CAS_RETRY_LIMIT {
+                        return Err(crate::grpc::persistence_error_to_status(err, operation));
+                    }
+                    debug!(
+                        sandbox_id,
+                        attempt,
+                        current_resource_version,
+                        "Sandbox delete phase transition conflicted; retrying"
+                    );
+                    tokio::task::yield_now().await;
+                }
+                Err(err) => return Err(crate::grpc::persistence_error_to_status(err, operation)),
+            }
+        }
+
+        unreachable!("delete phase retry loop always returns")
+    }
+
+    async fn write_sandbox_phase_deleting_from_snapshot(
+        &self,
+        mut sandbox: Sandbox,
+    ) -> crate::persistence::PersistenceResult<Sandbox> {
+        let id = sandbox.object_id().to_string();
+        let name = sandbox.object_name().to_string();
+        let expected_resource_version = sandbox
+            .metadata
+            .as_ref()
+            .map_or(0, |metadata| metadata.resource_version);
+
+        sandbox.set_phase(SandboxPhase::Deleting as i32);
+
+        let labels_json = sandbox
+            .metadata
+            .as_ref()
+            .map(|metadata| &metadata.labels)
+            .filter(|labels| !labels.is_empty())
+            .map(serde_json::to_string)
+            .transpose()
+            .map_err(|e| {
+                crate::persistence::PersistenceError::Encode(format!(
+                    "failed to serialize labels: {e}"
+                ))
+            })?;
+
+        let result = self
+            .store
+            .put_if(
+                Sandbox::object_type(),
+                &id,
+                &name,
+                &sandbox.encode_to_vec(),
+                labels_json.as_deref(),
+                WriteCondition::MatchResourceVersion(expected_resource_version),
+            )
+            .await?;
+
+        if let Some(metadata) = sandbox.metadata.as_mut() {
+            metadata.resource_version = result.resource_version;
+        }
+
+        Ok(sandbox)
     }
 
     pub fn spawn_watchers(&self, shutdown_rx: watch::Receiver<bool>) {
@@ -2534,6 +2634,58 @@ mod tests {
             compute_error_from_status(Status::failed_precondition("sandbox agent pod IP is not available")),
             ComputeError::Precondition(message) if message == "sandbox agent pod IP is not available"
         ));
+    }
+
+    #[tokio::test]
+    async fn set_sandbox_phase_deleting_retries_after_stale_snapshot_conflict() {
+        let runtime = test_runtime(Arc::new(TestDriver::default())).await;
+        let sandbox = sandbox_record("sb-1", "sandbox-a", SandboxPhase::Ready);
+        runtime.store.put_message(&sandbox).await.unwrap();
+
+        let stale_snapshot = runtime
+            .store
+            .get_message::<Sandbox>("sb-1")
+            .await
+            .unwrap()
+            .unwrap();
+
+        runtime
+            .store
+            .update_message_cas::<Sandbox, _>("sb-1", 0, |sandbox| {
+                sandbox.set_current_policy_version(7);
+            })
+            .await
+            .unwrap();
+
+        let updated = runtime
+            .set_sandbox_phase_deleting_with_initial_snapshot("sb-1", Some(stale_snapshot))
+            .await
+            .unwrap();
+
+        assert_eq!(
+            SandboxPhase::try_from(updated.phase()).unwrap(),
+            SandboxPhase::Deleting
+        );
+        assert_eq!(updated.current_policy_version(), 7);
+        assert_eq!(
+            updated
+                .metadata
+                .as_ref()
+                .map_or(0, |metadata| metadata.resource_version),
+            3
+        );
+
+        let stored = runtime
+            .store
+            .get_message::<Sandbox>("sb-1")
+            .await
+            .unwrap()
+            .unwrap();
+        assert_eq!(
+            SandboxPhase::try_from(stored.phase()).unwrap(),
+            SandboxPhase::Deleting
+        );
+        assert_eq!(stored.current_policy_version(), 7);
     }
 
     #[tokio::test]

